### PR TITLE
feat(web,server): user avatar color

### PR DESCRIPTION
**Pull request by martabal**
## Changes made in this PR

This PR adds the ability to a user to change its avatar color. Once he has set the avatar color, if he had a profile image, this profile image is deleted. Now, when a user is created, its avatar color is picked randomly from 10 colors.

Now when the user set a new profile image, the previous profile image is deleted from the file system.

Resolves #4767.

## Screenshots

https://github.com/immich-app/immich/assets/74269598/3556968c-3f65-4e33-962c-3a7f98f75808

### Responsive design

image

### CODE REVIEW

File: server/immich-openapi-specs.json [hunk 1]
```diff
@@ -5440,6 +5440,29 @@
       }
     },
     "/user/profile-image": {
+      "delete": {
+        "operationId": "deleteProfileImage",
+        "parameters": [],
+        "responses": {
+          "200": {
```
Comment (jrasm91) [inline server/immich-openapi-specs.json:5447]: This should be 204 no content

File: server/src/domain/user/dto/update-user.dto.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { ApiProperty } from '@nestjs/swagger';
 import { Transform } from 'class-transformer';
 import { IsBoolean, IsEmail, IsNotEmpty, IsString, IsUUID } from 'class-validator';
+import { UserAvatarColor } from '..';
```
Comment (jrasm91) [inline server/src/domain/user/dto/update-user.dto.ts:4]: Can you fix this import?

File: server/src/infra/entities/user.entity.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { UserAvatarColor } from '@app/domain/user/response-dto/user-response.dto';
```
Comment (jrasm91) [inline server/src/infra/entities/user.entity.ts:1]: Importing from app/domain into user entities can cause circular imports. Can you move the enum to this file instead?

File: web/src/lib/components/user-settings-page/avatar-selector.svelte [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  import CircleIconButton from '../elements/buttons/circle-icon-button.svelte';
+  import { createEventDispatcher } from 'svelte';
+  import FullScreenModal from '../shared-components/full-screen-modal.svelte';
+  import UserAvatar from '../shared-components/user-avatar.svelte';
+  import { UpdateUserDtoAvatarColorEnum, UserResponseDto } from '@api';
+
+  export let user: UserResponseDto;
+
+  const dispatch = createEventDispatcher();
+  const colors: Array<UpdateUserDtoAvatarColorEnum> = [
```
Comment (jrasm91) [inline web/src/lib/components/user-settings-page/avatar-selector.svelte:12]: Can this just be Object.values(UpdateUserDtoAvatarColorEnum)? Otherwise you have to remember to update the list whenever it changes.

File: server/src/domain/user/dto/update-user.dto.ts [hunk 5]
```diff
@@ -49,4 +50,13 @@ export class UpdateUserDto {
   @Optional()
   @IsBoolean()
   memoriesEnabled?: boolean;
+
+  @Optional()
+  @IsString()
+  @IsNotEmpty()
+  avatarColor?: UserAvatarColor;
```
Comment (jrasm91) [inline server/src/domain/user/dto/update-user.dto.ts:57]: This should not be `IsString()` but `IsEnum(UseAvatarColor)`, right? Also can you add the `@ApiProperty()` stuff for named enums, like we do everywhere else?

File: server/src/domain/user/user.service.ts [hunk 6]
```diff
@@ -93,10 +93,25 @@ export class UserService {
...
     const updatedUser = await this.userRepository.update(authUser.id, { profileImagePath: fileInfo.path });
+    if (user.profileImagePath !== '') {
+      const files = [user.profileImagePath];
+      await this.jobRepository.queue({ name: JobName.DELETE_FILES, data: { files } });
+    }
     return mapCreateProfileImageResponse(updatedUser.id, updatedUser.profileImagePath);
   }
 
+  async deleteProfileImage(authUser: AuthUserDto): Promise<void> {
+    const user = await this.findOrFail(authUser.id, { withDeleted: true });
```
Comment (jrasm91) [inline server/src/domain/user/user.service.ts:106]: You can delete a deleted user's profile pictures? Seems a bit strange.

File: server/src/domain/user/user.service.ts [hunk 7]
```diff
@@ -93,10 +93,25 @@ export class UserService {
     authUser: AuthUserDto,
     fileInfo: Express.Multer.File,
   ): Promise<CreateProfileImageResponseDto> {
+    const user = await this.findOrFail(authUser.id, { withDeleted: true });
     const updatedUser = await this.userRepository.update(authUser.id, { profileImagePath: fileInfo.path });
+    if (user.profileImagePath !== '') {
```
Comment (jrasm91) [inline server/src/domain/user/user.service.ts:98]: How or why would this ever be an empty string when using "createProfileImage"?

File: server/src/domain/user/dto/update-user.dto.ts [hunk 8]
```diff
@@ -49,4 +50,13 @@ export class UpdateUserDto {
...
+
+  @Optional()
+  @IsString()
+  profileImagePath?: string;
```
Comment (jrasm91) [inline server/src/domain/user/dto/update-user.dto.ts:61]: Why is this added here?

File: web/src/lib/components/shared-components/user-avatar.svelte [hunk 9]
```diff
@@ -39,12 +45,14 @@
     md: 'w-10 h-10',
     lg: 'w-12 h-12',
     xl: 'w-20 h-20',
+    xxl: 'w-24 h-24',
+    xxxl: 'w-24 h-24',
```
Comment (jrasm91) [inline web/src/lib/components/shared-components/user-avatar.svelte:49]: are these supposed to be the same size?

File: server/src/domain/user/user.service.ts (hunk 7 above)
Comment (martabal) [reply server/src/domain/user/user.service.ts:98]: If a user has never set is profile image, the default value is an empty string

Comment (jrasm91) [reply server/src/domain/user/user.service.ts:98]: Oh, I see. What do you think about having it be a static path instead of a random uuid and then having to delete the old one? So overwrite instead of new + delete.

File: server/src/domain/user/dto/update-user.dto.ts (hunk 8 above)
Comment (martabal) [reply server/src/domain/user/dto/update-user.dto.ts]: A rest of my previous implementation, I have reverted it

File: server/src/domain/user/user.service.ts (hunk 7 above)
Comment (martabal) [reply server/src/domain/user/user.service.ts:98]: I don't follow. How do you determine if a user has set its profile picture ?

Comment (martabal) [reply server/src/domain/user/user.service.ts:98]: Yeah it should work. Should we use the user id for the static path ?

Comment (martabal) [reply server/src/domain/user/user.service.ts:98]: Maybe we should change the profileImagePath to be a boolean

Comment (jrasm91) [reply server/src/domain/user/user.service.ts:98]: Ah, hold on. I guess the extension may not always be the same, so you will have to potentially delete it afterwards, so maybe it is fine like you have it.

Comment (jrasm91) [reply server/src/domain/user/user.service.ts:98]: I did get confused with user vs updated user and the user.profilePath is the old path, but referenced after you updated to the new one. Maybe you could do something like `const { profilePath: oldPath } = await this.findOrFail(...)`; to make it a little more clear.

Comment (martabal) [reply server/src/domain/user/user.service.ts]: Done

**Comment by alextran1502** [comment]
Hello, thank you for working on this mechanism! 

I think we can improve the slow of not clicking many buttons by adding the edit profile image in the user profile picture from the popup panel. What do you think?

image

**Comment by martabal** [comment]
It looks good. Should it be in addition or replacement ?

**Comment by alextran1502** [comment]
> It looks good. Should it be in addition or replacement ?

I would say a replacement

**Comment by martabal** [comment]
Done 

https://github.com/immich-app/immich/assets/74269598/2a133113-190c-49b6-8dea-2ddf113507c8

**Comment by shenlong-tanwen** [comment]
Pushed the mobile changes to support displaying the avatar colors assigned from the web. It is still not possible to set a color from the mobile app, but that can be added in a separate PR.

>  [!Warning]
> The mobile app and server has to be in the same version. Only updating the mobile app will prevent the user from logging in due to changes in the dto.

File: server/src/domain/user/response-dto/user-response.dto.ts [hunk 10]
```diff
@@ -1,10 +1,17 @@
...
+  const randomIndex = Math.floor(Math.random() * values.length);
+  return values[randomIndex] as UserAvatarColor;
+};
 
 export class UserDto {
   id!: string;
   name!: string;
   email!: string;
   profileImagePath!: string;
+  avatarColor!: UserAvatarColor;
```
Comment (jrasm91) [inline server/src/domain/user/response-dto/user-response.dto.ts:14]: This is missing api property decorator for a named enum.

File: server/src/domain/user/response-dto/user-response.dto.ts [hunk 11]
```diff
@@ -1,10 +1,17 @@
-import { UserEntity } from '@app/infra/entities';
+import { UserAvatarColor, UserEntity } from '@app/infra/entities';
+
+export const getRandomAvatarColor = (): UserAvatarColor => {
+  const values = Object.values(UserAvatarColor);
+  const randomIndex = Math.floor(Math.random() * values.length);
+  return values[randomIndex] as UserAvatarColor;
+};
```
Comment (jrasm91) [inline server/src/domain/user/response-dto/user-response.dto.ts:7]: Why don't we just make this deterministic? I don't think need need to store it or anything like that. Empty implies random and if it was deterministic this could be accomplished in the mapper itself.

File: server/src/domain/user/user.core.ts [hunk 12]
```diff
@@ -98,7 +99,7 @@ export class UserCore {
     if (payload.storageLabel) {
       payload.storageLabel = sanitize(payload.storageLabel);
     }
-
+    payload.avatarColor = getRandomAvatarColor();
```
Comment (jrasm91) [inline server/src/domain/user/user.core.ts:102]: I would prefer making this deterministic without having to explicitly save it in the database on user create.

File: web/src/lib/components/shared-components/navigation-bar/avatar-selector.svelte [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  import { createEventDispatcher } from 'svelte';
+  import { UserDtoAvatarColorEnum, UserResponseDto } from '@api';
+  import CircleIconButton from '$lib/components/elements/buttons/circle-icon-button.svelte';
+  import FullScreenModal from '../full-screen-modal.svelte';
+  import UserAvatar from '../user-avatar.svelte';
+
+  export let user: UserResponseDto;
+
+  const dispatch = createEventDispatcher();
+  const colors: Array<UserDtoAvatarColorEnum> = Object.values(UserDtoAvatarColorEnum);
```
Comment (jrasm91) [inline web/src/lib/components/shared-components/navigation-bar/avatar-selector.svelte:12]:
```suggestion
  const colors: UserDtoAvatarColorEnum[] = Object.values(UserDtoAvatarColorEnum);
```

**Review by jrasm91 — APPROVED** [review]
Looks pretty good!

File: server/src/domain/user/response-dto/user-response.dto.ts (hunk 11 above)
Comment (martabal) [reply server/src/domain/user/response-dto/user-response.dto.ts:7]: I don't follow. The goal is to create a random avatar color only on user creation.

File: server/src/domain/user/response-dto/user-response.dto.ts [hunk 14]
```diff
@@ -1,10 +1,28 @@
...
 
 export class UserDto {
   id!: string;
   name!: string;
   email!: string;
   profileImagePath!: string;
+  @IsEnum(UserAvatarColor)
+  @Optional()
+  @ApiProperty({ enumName: 'UserAvatarColor', enum: UserAvatarColor })
+  avatarColor!: UserAvatarColor | null;
```
Comment (jrasm91) [inline server/src/domain/user/response-dto/user-response.dto.ts:25]: Under what condition will this be null? Line 46 says "if null pick a value". Is this just to avoid some mobile compat issue?

File: server/src/domain/user/user.service.ts [hunk 15]
```diff
@@ -93,10 +93,25 @@ export class UserService {
     authUser: AuthUserDto,
     fileInfo: Express.Multer.File,
   ): Promise<CreateProfileImageResponseDto> {
+    const { profileImagePath: oldpath } = await this.findOrFail(authUser.id, { withDeleted: false });
     const updatedUser = await this.userRepository.update(authUser.id, { profileImagePath: fileInfo.path });
+    if (oldpath !== '') {
+      const files = [oldpath];
+      await this.jobRepository.queue({ name: JobName.DELETE_FILES, data: { files } });
```
Comment (jrasm91) [inline server/src/domain/user/user.service.ts:100]:
```suggestion
      await this.jobRepository.queue({ name: JobName.DELETE_FILES, data: { files: [oldPath] } });
```

File: server/src/domain/user/user.service.ts [hunk 16]
```diff
@@ -93,10 +93,25 @@ export class UserService {
...
   }
 
+  async deleteProfileImage(authUser: AuthUserDto): Promise<void> {
+    const user = await this.findOrFail(authUser.id, { withDeleted: false });
+    if (user.profileImagePath !== '') {
+      await this.userRepository.update(authUser.id, { profileImagePath: '' });
+      const files = [user.profileImagePath];
+      await this.jobRepository.queue({ name: JobName.DELETE_FILES, data: { files } });
+    }
+    return;
```
Comment (jrasm91) [inline server/src/domain/user/user.service.ts:112]: I don't think this is necessary, is it?

File: server/src/domain/user/user.service.ts [hunk 17]
```diff
@@ -93,10 +93,25 @@ export class UserService {
...
+    if (oldpath !== '') {
+      const files = [oldpath];
+      await this.jobRepository.queue({ name: JobName.DELETE_FILES, data: { files } });
+    }
     return mapCreateProfileImageResponse(updatedUser.id, updatedUser.profileImagePath);
   }
 
+  async deleteProfileImage(authUser: AuthUserDto): Promise<void> {
+    const user = await this.findOrFail(authUser.id, { withDeleted: false });
+    if (user.profileImagePath !== '') {
```
Comment (jrasm91) [inline server/src/domain/user/user.service.ts:107]: I think an http error make be more appropriate if the user doesn't have a profile picture. The client can ignore it instead.

**Review by jrasm91 — APPROVED** [review]
Looks pretty good!

File: server/src/domain/user/response-dto/user-response.dto.ts (hunk 14 above)
Comment (martabal) [reply server/src/domain/user/response-dto/user-response.dto.ts:25]: Else there's a type error with `const providedPassword = await ask(admin);` https://github.com/immich-app/immich/blob/ac7e8bcdf469bfbe9810c8ccfc39ebb1e6b37b45/server/src/domain/user/user.service.ts#L108-L120

Comment (jrasm91) [reply server/src/domain/user/response-dto/user-response.dto.ts:25]: That would be because it is returning an user entity not a user response. Can you change it to `await ask(mapUser(admin))` instead? And then remove the null type on the response dto.

Comment (martabal) [reply server/src/domain/user/response-dto/user-response.dto.ts]: Damn, I didn't think about this trick. Thanks !
